lib/messages/chainlock.js

```diff
@@ -0,0 +1,276 @@
+      var thresholdSignature = instance.Signature.fromBytes(Buffer.from(signature,'hex'));
+
+      thresholdSignature.setAggregationInfo(aggregationInfo);
+
+      var isValid = thresholdSignature.verify();
+
+      return resolve(isValid);
+    })
+  });
+}
+
+/**
+ * Validate Chainlock structure
+ */
+ChainLock.prototype.validate = function validate() {
+  $.checkArgument(utils.isUnsignedInteger(this.height), 'Expect height to be an unsigned integer');
+  $.checkArgument(isHexStringOfSize(this.blockHash.toString('hex'), SHA256_HASH_SIZE * 2), `Expected blockhash to be a hex string of size ${SHA256_HASH_SIZE}`);
+  $.checkArgument(isHexStringOfSize(this.signature.toString('hex'), BLS_SIGNATURE_SIZE * 2), 'Expected signature to be a bls signature');
+};
+
+/**
+ * Returns chainLock hash
+ * @returns {Buffer}
+ */
+ChainLock.prototype.getHash = function getHash() {
+  return doubleSha256(this.toBuffer()).reverse();
+};
+
+/**
+ * Computes the request ID for this ChainLock
+ * @returns {Buffer} - Request id for this chainlock
+ */
+ChainLock.prototype.getRequestId = function getRequestId() {
+  var bufferWriter = new BufferWriter();
+
+  var prefix = ChainLock.CLSIG_REQUESTID_PREFIX;
+  var prefixLength = prefix.length;
+
+  bufferWriter.writeVarintNum(prefixLength);
+  bufferWriter.write(Buffer.from(ChainLock.CLSIG_REQUESTID_PREFIX, 'utf-8'));
+  bufferWriter.writeUInt32LE(this.height);
+
+  // Double-sha is used to protect from extension attacks.
+  return doubleSha256(bufferWriter.toBuffer());
+};
+
+/**
+ * Computes signature id for a quorum entry
+ * @param {QuorumEntry} quorumEntry
+ * @returns {Buffer} - Signature id for this requestId and quorum.
+ */
+ChainLock.prototype.getSignatureIDForQuorumEntry = function (quorumEntry) {
+  var llmqType = quorumEntry.llmqType;
+  var quorumHash = quorumEntry.quorumHash;
+  var requestID = this.getRequestId();
+  var blockHash = this.blockHash;
+
+  var bufferWriter = new BufferWriter();
+  bufferWriter.writeUInt8(llmqType);
+  bufferWriter.write(Buffer.from(quorumHash, 'hex'));
+  bufferWriter.write(Buffer.from(requestID, 'hex'));
+  bufferWriter.write(blockHash);
+  return doubleSha256(bufferWriter.toBuffer());
+}
+
+/**
+ * Serializes chainlock to JSON
+ * @returns {Object} A plain object with the chainlock information
+ */
+ChainLock.prototype.toJSON = ChainLock.prototype.toObject = function toJSON() {
+  return {
+    height: this.height,
+    blockHash: BufferUtil.reverse(this.blockHash).toString('hex'),
+    signature: this.signature.toString('hex')
+  };
+};
+
+
+/**
+ * Serialize ChainLock
+ * @returns {string} - A hex encoded string of the chainlock
+ */
+ChainLock.prototype.toString = function toString() {
+  return this.toBuffer().toString('hex');
+};
+
+/**
+ * Serialize ChainLock to buffer
+ * @return {Buffer}
+ */
+ChainLock.prototype.toBuffer = function toBuffer() {
+  return this.toBufferWriter().concat();
```

Comment:
Why not to use `.toBufferWriter().toBuffer()`? I think it would be a little bit more readable this way

Reply:
Good eyes ! Also modified some other similar elements (block, merkleblock, blockheader...)